pkg/jobstore/types.go

```diff
@@ -5,18 +5,31 @@ import (
 	"context"
 
 	"github.com/bacalhau-project/bacalhau/pkg/models"
+	"k8s.io/apimachinery/pkg/labels"
 )
 
 type JobQuery struct {
-	ID          string   `json:"id"`
-	Namespace   string   `json:"namespace"`
-	IncludeTags []string `json:"include_tags"`
-	ExcludeTags []string `json:"exclude_tags"`
-	Limit       uint32   `json:"limit"`
-	Offset      uint32   `json:"offset"`
-	ReturnAll   bool     `json:"return_all"`
-	SortBy      string   `json:"sort_by"`
-	SortReverse bool     `json:"sort_reverse"`
+	ID        string
+	Namespace string
+
+	// IncludeTags and ExcludeTags are used primarily by the requester's list API.
+	// In the orchestrator API, we insted use the Selector field to filter jobs.
+	IncludeTags []string
+	ExcludeTags []string
+	Limit       uint32
+	Offset      uint32
+	ReturnAll   bool
+	SortBy      string
+	SortReverse bool
+	Selector    labels.Selector
+}
+
+type JobQueryResponse struct {
+	Jobs            []models.Job
+	Offset          uint32 // Offset into the filtered results of the first returned record
+	Limit           uint32 // The number of records to return, 0 means all
+	NextOffset      uint32 // Offset + Limit of the next page of results, 0 means no more results
+	RecordsReturned uint32 // The number of jobs actually held in the Jobs field
```

Comment:
The way I see it, every data store implementation will have its own way to handle pagination and what it fields it requires to efficiently fetch the next set of results.

For that reason I recommend that `NextToken` to be a base64 string that is generated and decoded by the jobstore implementations. In boltdb, it can be an `offset` based on current implementation, along with sorting criteria to make sure the user is getting consistent results. We can add more fields in the future, or using different mechanism than offsets to gurantee consistent paginations.

If the user sends a follow up request with a sorting criteria doesn't match the one embedded in the token, then we can fail the job or ignore the one in the request.

Future implementations might encode a cursor reference and the requester node that handled the original request and has a cursor open to handle subsequent requests.

What do you think?

Reply:
Yeah sounds good, search criteria, limit and offset in a base64 encoded blob. It _might_ get a bit unwieldy length wise, but perhaps we can more simply encode it without the verbosity of JSON (even | separated + base64 might be okay).